File: src/scanner/express.js
import path from "node:path";
import { parse } from "@babel/parser";
import _traverse from "@babel/traverse";
import { confidenceForFile } from "./confidence.js";
const traverse = _traverse.default;

/**
 * AST-based extractor - replaces the regex heuristic in the previous version.
 * Same fact schema as before (id, type, method, routePath, receiver, file,
 * line, evidence), plus a v3 `confidence` field (see confidence.js).
 * Extraction method changed; nothing else did.
 *
 * Still matches the same receiver names as the regex version ("app" /
 * "router") rather than tracing express() / express.Router() call sites
 * to their variable bindings - a natural follow-up if we want to
 * catch renamed routers (e.g. userRouter.get(...)), but out of scope here
 * to keep behavior parity with v1.
 */

const HTTP_METHODS = new Set(["get", "post", "put", "delete", "patch", "use", "all"]);
const RECEIVER_NAMES = new Set(["app", "router"]);
const EVIDENCE_MAX_CHARS = 160;

function evidenceFor(lines, ln) {
  return lines[ln - 1]?.trim().slice(0, EVIDENCE_MAX_CHARS) || "";
}

// Resolve a route-path argument to a plain string. Handles string literals
// and no-substitution template literals (the AST equivalent of the regex's
// ['"`] character class); anything dynamic (template with expressions,
// identifiers, concatenation) is not a route we can name statically.
function staticStringValue(argNode) {
  if (!argNode) return null;
  if (argNode.type === "StringLiteral") return argNode.value;
  if (argNode.type === "TemplateLiteral" && argNode.expressions.length === 0) {
    return argNode.quasis.map((q) => q.value.cooked).join("");
  }
  return null;
}

export function extractExpressRoutes(filePath, content, rootDir, nextId) {
  const relPath = path.relative(rootDir, filePath);
  const confidence = confidenceForFile(relPath);
  const facts = [];
  const lines = content.split("\n");

  let ast;
  try {
    ast = parse(content, {
      sourceType: "unambiguous",
      plugins: ["jsx", "typescript", "classProperties", "decorators-legacy"],
      errorRecovery: true,
    });
  } catch (err) {
    // Unparseable file - skip silently, same as v1 on garbage input.
    return facts;
  }

  // See facts.js for why traverse() gets its own try/catch: a valid AST can
  // still make traverse() throw on unusual real-world code shapes, and one
  // bad file shouldn't kill the whole repo scan.
  try {
    traverse(ast, {
      CallExpression(path) {
        const callee = path.node.callee;
        if (callee.type !== "MemberExpression" || callee.computed) return;
        if (callee.object.type !== "Identifier") return;
        if (callee.property.type !== "Identifier") return;

        const receiver = callee.object.name;
        const method = callee.property.name;
        if (!RECEIVER_NAMES.has(receiver) || !HTTP_METHODS.has(method)) return;

        const routePath = staticStringValue(path.node.arguments[0]);
        if (routePath === null) return;
        if (method === "use" && !routePath.startsWith("/")) return;

        const ln = path.node.loc?.start.line;
        facts.push({
          id: nextId("route"),
          type: "express_route",
          method: method.toUpperCase(),
          routePath,
          receiver,
          file: relPath,
          line: ln,
          confidence,
          evidence: evidenceFor(lines, ln),
        });
      },
    });
  } catch (err) {
    // return partial facts
  }

  return facts;
}

File: python/src/rune/scanner/express.py
import os
import re

from .lines import build_line_index, line_number_at

ROUTE_RE = re.compile(r"""\b(app|router)\.(get|post|put|delete|patch|use|all)\(\s*['"`]([^'"`]+)['"`]""")


def extract_express_routes(file_path: str, content: str, root_dir: str, next_id) -> list[dict]:
    """
    Extracts Express route facts from a file's raw content.
    Heuristic: matches `app.<method>('/path', ...)` and `router.<method>('/path', ...)`.
    Does not resolve router mount prefixes across files in v1 (documented limitation).
    """
    rel_path = os.path.relpath(file_path, root_dir)
    facts: list[dict] = []
    lines = content.split("\n")
    line_offsets = build_line_index(content)

    for m in ROUTE_RE.finditer(content):
        receiver, method, route_path = m.group(1), m.group(2), m.group(3)
        if method == "use" and not route_path.startswith("/"):
            continue
        ln = line_number_at(line_offsets, m.start())
        evidence = lines[ln - 1].strip()[:160] if 1 <= ln <= len(lines) else ""
        facts.append({
            "id": next_id("route"),
            "type": "express_route",
            "method": method.upper(),
            "routePath": route_path,
            "receiver": receiver,
            "file": rel_path,
            "line": ln,
            "evidence": evidence,
        })
    return facts
